File: backend/src/main/java/com/schoolfinder/service/AdminService.java
package com.schoolfinder.service;

import com.schoolfinder.api.Dtos.ImageDto;
import com.schoolfinder.api.Dtos.PageResponse;
import com.schoolfinder.api.Dtos.ReviewDto;
import com.schoolfinder.api.Dtos.SchoolDetail;
import com.schoolfinder.api.Dtos.SchoolUpsertRequest;
import com.schoolfinder.api.Dtos.UserDto;
import com.schoolfinder.api.NotFoundException;
import com.schoolfinder.domain.AppUser;
import com.schoolfinder.domain.Review;
import com.schoolfinder.domain.ReviewStatus;
import com.schoolfinder.domain.Role;
import com.schoolfinder.domain.School;
import com.schoolfinder.domain.SchoolImage;
import com.schoolfinder.repository.AppUserRepository;
import com.schoolfinder.repository.ReviewRepository;
import com.schoolfinder.repository.SchoolImageRepository;
import com.schoolfinder.repository.SchoolRepository;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AdminService {

    private final SchoolRepository schools;
    private final AppUserRepository users;
    private final ReviewRepository reviews;
    private final SchoolImageRepository images;
    private final CloudinaryService cloudinary;
    private final SchoolService schoolService;

    public AdminService(
        SchoolRepository schools,
        AppUserRepository users,
        ReviewRepository reviews,
        SchoolImageRepository images,
        CloudinaryService cloudinary,
        SchoolService schoolService
    ) {
        this.schools = schools;
        this.users = users;
        this.reviews = reviews;
        this.images = images;
        this.cloudinary = cloudinary;
        this.schoolService = schoolService;
    }

    // ----- School images (Cloudinary) -----

    @Transactional
    public ImageDto uploadSchoolImage(Long id, byte[] bytes, String caption, boolean setCover) {
        School school = schools.findById(id).orElseThrow(() -> new NotFoundException("School " + id + " not found"));
        String url = cloudinary.upload(bytes);
        SchoolImage img = images.save(new SchoolImage(school, url, caption));
        if (setCover) {
            school.setCoverImageUrl(url);
            schools.save(school);
        }
        return new ImageDto(img.getId(), img.getUrl(), img.getCaption());
    }

    @Transactional
    public void deleteSchoolImage(Long imageId) {
        if (!images.existsById(imageId)) throw new NotFoundException("Image " + imageId + " not found");
        images.deleteById(imageId);
    }

    // ----- Schools -----

    @Transactional
    public SchoolDetail createSchool(SchoolUpsertRequest req) {
        School school = new School(req.name(), req.category());
        applyFrom(school, req);
        return schoolService.detail(schools.save(school).getId());
    }

    @Transactional
    public SchoolDetail updateSchool(Long id, SchoolUpsertRequest req) {
        School school = schools.findById(id).orElseThrow(() -> new NotFoundException("School " + id + " not found"));
        school.setName(req.name());
        school.setCategory(req.category());
        applyFrom(school, req);
        schools.save(school);
        return schoolService.detail(id);
    }

    @Transactional
    public void deleteSchool(Long id) {
        if (!schools.existsById(id)) throw new NotFoundException("School " + id + " not found");
        schools.deleteById(id);
    }

    private void applyFrom(School school, SchoolUpsertRequest req) {
        school.setDescription(req.description());
        school.setHistory(req.history());
        school.setCity(req.city());
        school.setRegion(req.region());
        school.setAddress(req.address());
        school.setLatitude(req.latitude());
        school.setLongitude(req.longitude());
        school.setTuitionFee(req.tuitionFee());
        school.setCurrency(req.currency());
        school.setWebsite(req.website());
        school.setPhone(req.phone());
        school.setEmail(req.email());
        school.setCoverImageUrl(req.coverImageUrl());
        school.setApproved(req.approved());
    }

    // ----- Users -----

    @Transactional(readOnly = true)
    public List<UserDto> listUsers() {
        return users.findAll().stream()
            .map(u -> new UserDto(u.getId(), u.getEmail(), u.getDisplayName(), u.getRole(), u.isActive(), u.getCreatedAt()))
            .toList();
    }

    @Transactional
    public UserDto setUserActive(Long id, boolean active) {
        AppUser u = users.findById(id).orElseThrow(() -> new NotFoundException("User " + id + " not found"));
        u.setActive(active);
        return new UserDto(u.getId(), u.getEmail(), u.getDisplayName(), u.getRole(), u.isActive(), u.getCreatedAt());
    }

    @Transactional
    public UserDto setUserRole(Long id, Role role) {
        AppUser u = users.findById(id).orElseThrow(() -> new NotFoundException("User " + id + " not found"));
        u.setRole(role);
        return new UserDto(u.getId(), u.getEmail(), u.getDisplayName(), u.getRole(), u.isActive(), u.getCreatedAt());
    }

    // ----- Review moderation -----

    @Transactional(readOnly = true)
    public PageResponse<ReviewDto> pendingReviews(Pageable pageable) {
        return PageResponse.of(
            reviews.findByStatusOrderByCreatedAtDesc(ReviewStatus.PENDING, pageable).map(this::toReviewDto));
    }

    @Transactional
    public ReviewDto moderateReview(Long id, ReviewStatus status) {
        Review review = reviews.findById(id).orElseThrow(() -> new NotFoundException("Review " + id + " not found"));
        review.setStatus(status);
        return toReviewDto(review);
    }

    // ----- Analytics -----

    @Transactional(readOnly = true)
    public Map<String, Object> analytics() {
        Map<String, Long> byCategory = new TreeMap<>();
        for (School s : schools.findAll()) {
            byCategory.merge(s.getCategory().name(), 1L, Long::sum);
        }
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("totalSchools", schools.count());
        result.put("totalUsers", users.count());
        result.put("totalReviews", reviews.count());
        result.put("schoolsByCategory", byCategory);
        return result;
    }

    private ReviewDto toReviewDto(Review r) {
        String displayName = r.getUser().getDisplayName() != null ? r.getUser().getDisplayName() : r.getUser().getEmail();
        return new ReviewDto(
            r.getId(),
            r.getSchool().getId(),
            r.getSchool().getName(),
            displayName,
            r.getRating(),
            r.getComment(),
            r.getStatus(),
            r.getCreatedAt()
        );
    }
}

File: backend/src/main/kotlin/com/schoolfinder/service/AdminService.kt
package com.schoolfinder.service

import com.schoolfinder.api.ImageDto
import com.schoolfinder.api.NotFoundException
import com.schoolfinder.api.PageResponse
import com.schoolfinder.api.ReviewDto
import com.schoolfinder.api.SchoolDetail
import com.schoolfinder.api.SchoolUpsertRequest
import com.schoolfinder.api.UserDto
import com.schoolfinder.domain.Review
import com.schoolfinder.domain.ReviewStatus
import com.schoolfinder.domain.Role
import com.schoolfinder.domain.School
import com.schoolfinder.domain.SchoolImage
import com.schoolfinder.repository.AppUserRepository
import com.schoolfinder.repository.ReviewRepository
import com.schoolfinder.repository.SchoolImageRepository
import com.schoolfinder.repository.SchoolRepository
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AdminService(
    private val schools: SchoolRepository,
    private val users: AppUserRepository,
    private val reviews: ReviewRepository,
    private val images: SchoolImageRepository,
    private val cloudinary: CloudinaryService,
    private val schoolService: SchoolService,
) {

    // ----- School images (Cloudinary) -----

    @Transactional
    fun uploadSchoolImage(id: Long, bytes: ByteArray, caption: String?, setCover: Boolean): ImageDto {
        val school = schools.findById(id).orElseThrow { NotFoundException("School $id not found") }
        val url = cloudinary.upload(bytes)
        val img = images.save(SchoolImage(school = school, url = url, caption = caption))
        if (setCover) {
            school.coverImageUrl = url
            schools.save(school)
        }
        return ImageDto(img.id!!, img.url, img.caption)
    }

    @Transactional
    fun deleteSchoolImage(imageId: Long) {
        if (!images.existsById(imageId)) throw NotFoundException("Image $imageId not found")
        images.deleteById(imageId)
    }

    // ----- Schools -----

    @Transactional
    fun createSchool(req: SchoolUpsertRequest): SchoolDetail {
        val school = School(name = req.name, category = req.category).apply { applyFrom(req) }
        return schoolService.detail(schools.save(school).id!!)
    }

    @Transactional
    fun updateSchool(id: Long, req: SchoolUpsertRequest): SchoolDetail {
        val school = schools.findById(id).orElseThrow { NotFoundException("School $id not found") }
        school.name = req.name
        school.category = req.category
        school.applyFrom(req)
        schools.save(school)
        return schoolService.detail(id)
    }

    @Transactional
    fun deleteSchool(id: Long) {
        if (!schools.existsById(id)) throw NotFoundException("School $id not found")
        schools.deleteById(id)
    }

    private fun School.applyFrom(req: SchoolUpsertRequest) {
        description = req.description
        city = req.city
        region = req.region
        address = req.address
        latitude = req.latitude
        longitude = req.longitude
        tuitionFee = req.tuitionFee
        currency = req.currency
        website = req.website
        phone = req.phone
        email = req.email
        coverImageUrl = req.coverImageUrl
        approved = req.approved
    }

    // ----- Users -----

    @Transactional(readOnly = true)
    fun listUsers(): List<UserDto> = users.findAll().map {
        UserDto(it.id!!, it.email, it.displayName, it.role, it.active, it.createdAt)
    }

    @Transactional
    fun setUserActive(id: Long, active: Boolean): UserDto {
        val u = users.findById(id).orElseThrow { NotFoundException("User $id not found") }
        u.active = active
        return UserDto(u.id!!, u.email, u.displayName, u.role, u.active, u.createdAt)
    }

    @Transactional
    fun setUserRole(id: Long, role: Role): UserDto {
        val u = users.findById(id).orElseThrow { NotFoundException("User $id not found") }
        u.role = role
        return UserDto(u.id!!, u.email, u.displayName, u.role, u.active, u.createdAt)
    }

    // ----- Review moderation -----

    @Transactional(readOnly = true)
    fun pendingReviews(pageable: Pageable): PageResponse<ReviewDto> =
        PageResponse.of(reviews.findByStatusOrderByCreatedAtDesc(ReviewStatus.PENDING, pageable).map(::toReviewDto))

    @Transactional
    fun moderateReview(id: Long, status: ReviewStatus): ReviewDto {
        val review = reviews.findById(id).orElseThrow { NotFoundException("Review $id not found") }
        review.status = status
        return toReviewDto(review)
    }

    // ----- Analytics -----

    @Transactional(readOnly = true)
    fun analytics(): Map<String, Any> = mapOf(
        "totalSchools" to schools.count(),
        "totalUsers" to users.count(),
        "totalReviews" to reviews.count(),
        "schoolsByCategory" to schools.findAll().groupingBy { it.category.name }.eachCount(),
    )

    private fun toReviewDto(r: Review) = ReviewDto(
        id = r.id!!,
        schoolId = r.school.id!!,
        schoolName = r.school.name,
        userDisplayName = r.user.displayName ?: r.user.email,
        rating = r.rating.toInt(),
        comment = r.comment,
        status = r.status,
        createdAt = r.createdAt,
    )
}
